### PR TITLE
feat: use global omni token in token deployer

### PR DESCRIPTION
**Pull request by frolvanya**
TODO:
- [x] Fix tests
- [x] Test how migration works with global contracts
- [x] Add a test to cover token deployer migration

### CODE REVIEW

File: near/mock/mock-global-contract-deployer/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+# the command may be any combination of flags of `cargo-near`,
+# supported by respective version of binary inside the container besides `--no-locked` flag
+container_build_command = ["cargo", "near", "build", "non-reproducible-wasm", "--locked"]
+
+[lib]
+crate-type = ["cdylib", "rlib"]
+
+[dependencies]
+near-sdk = { workspace = true, features = ["global-contracts"] }
+near-contract-standards.workspace = true
```
[the remark this answers is not in the document]
Comment (frolvanya) [reply near/mock/mock-global-contract-deployer/Cargo.toml]: Fixed in 68e64dd3ff4277245b802ce49962acfbd7453336

File: near/mock/mock-global-contract-deployer/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        Promise::new(account_id)
+            .create_account()
+            .transfer(env::attached_deposit())
+            .add_full_access_key(env::signer_account_pk())
+            .deploy_global_contract(code_bytes)
+    }
+
+    /// Deploy a global contract, identifiable by the predecessor's account ID
+    #[payable]
+    pub fn deploy_global_contract_by_account_id(
```
Comment (frolvanya) [inline near/mock/mock-global-contract-deployer/src/lib.rs]: This mock contract is a workaround. `workspaces-rs` doesn't have any API to deploy global contract, so I could either use mock contract or implement direct RPC calls using jsonrpc client. Both options were bad, but mock contract is easier to implement and use. Also this option doesn't add any new dependencies in `omni-tests` project

File: near/omni-bridge/src/tests/lib_test.rs [hunk 3]
```diff
@@ -656,7 +656,9 @@ fn test_fin_transfer_callback_invalid_proof() {
         vec![PromiseResult::Failed],
     );
 
-    contract.fin_transfer_callback(&storage_actions, predecessor);
+    contract
+        .fin_transfer_callback(&storage_actions, predecessor)
+        .detach();
```
Comment (frolvanya) [inline near/omni-bridge/src/tests/lib_test.rs]: FYI, in `near-sdk-rs v5.18.0` all "dangling" promises must have an explicit `detach`. This was added to reduce number of bugs related to unreturned promises by accident

File: near/token-deployer/src/lib.rs [hunk 4]
```diff
@@ -50,18 +56,11 @@ impl TokenDeployer {
         contract
     }
 
-    #[payable]
     #[access_control_any(roles(Role::Controller))]
```
Comment (frolvanya) [inline near/token-deployer/src/lib.rs]: New tokens could be deployed for free, so we don't need any extra balance:

https://testnet.nearblocks.io/txns/E8ut6KdrXdvcsCw7Fac9xbr8tW4W2KGBvwrAohZMXhbX

File: near/omni-bridge/src/lib.rs [hunk 5]
```diff
@@ -1415,16 +1469,16 @@ impl Contract {
...
                 if self.deployed_tokens.contains(&token) {
-                    ext_token::ext(token).with_static_gas(MINT_TOKEN_GAS).mint(
-                        fee_recipient.clone(),
-                        transfer_message.fee.fee,
-                        None,
-                    );
+                    ext_token::ext(token)
+                        .with_static_gas(MINT_TOKEN_GAS)
+                        .mint(fee_recipient.clone(), transfer_message.fee.fee, None)
+                        .detach();
```
[the remark this answers is not in the document]
Comment (frolvanya) [reply near/omni-bridge/src/lib.rs]: Are you talking about `detach`? It was added in a new near-sdk version to reduce number of mistakes related to the detached promises:
https://github.com/near/near-sdk-rs/pull/1400

https://github.com/Near-One/omni-bridge/pull/473#discussion_r2570871518

File: near/token-deployer/src/lib.rs [hunk 6]
```diff
@@ -35,13 +33,21 @@ pub enum Role {
     duration_update_stagers(Role::DAO),
     duration_update_appliers(Role::DAO),
 ))]
-pub struct TokenDeployer {}
+pub struct TokenDeployer {
+    omni_token_global_contract_id: AccountId,
```
[the remark this answers is not in the document]
Comment (frolvanya) [reply near/token-deployer/src/lib.rs]: Added back in 520a80079caf543a61889bd7096e9e1d22a64bdc

File: near/omni-bridge/src/lib.rs [hunk 7]
```diff
@@ -2001,7 +2007,6 @@ impl Contract {
         require!(self.deployed_tokens.insert(&token_id), "ERR_TOKEN_EXIST");
         let required_deposit = env::storage_byte_cost()
             .saturating_mul((env::storage_usage().saturating_sub(storage_usage)).into())
-            .saturating_add(storage::BRIDGE_TOKEN_INIT_BALANCE)
```
[the remark this answers is not in the document]
Comment (kiseln) [reply near/omni-bridge/src/lib.rs]: When is one or the other used?

File: near/omni-tests/src/environment.rs [hunk 8]
```diff
@@ -76,6 +76,12 @@ impl TestEnvBuilder {
         )
         .await?;
 
+        self.worker
+            .root_account()?
+            .transfer_near(token_contract.id(), NearToken::from_near(1))
```
[the remark this answers is not in the document]
Comment (kiseln) [reply near/omni-tests/src/environment.rs]: Why do we need Near on token contracts?

File: near/token-deployer/src/migrate.rs
[the remark this answers is not in the document]
Comment (frolvanya) [reply near/token-deployer/src/migrate.rs]: I believe some tokens still have `factory.bridge.near` as their controller, so it'll be tedious to set a new controller to all this tokens

cc @karim-en

File: near/omni-token/src/lib.rs [hunk 9]
```diff
@@ -56,6 +62,8 @@ impl OmniToken {
             "Only the deployer account can init this contract"
         );
 
+        env::storage_write(IS_USING_GLOBAL_TOKEN_KEY, &[is_using_global_token.into()]);
```
[the remark this answers is not in the document]
Comment (frolvanya) [reply near/omni-token/src/lib.rs]: In order to avoid a new field in a state, since it will require a migration

cc @karim-en

File: near/omni-bridge/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (frolvanya) [reply near/omni-bridge/src/lib.rs]: We will solve this issue by upgrading factory.bridge.near to have a new deployer's code that would deploy tokens using global contract for _free_

File: near/omni-tests/src/environment.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (kiseln) [reply near/omni-tests/src/environment.rs]: Isn't it the other way around?

File: near/omni-token/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (kiseln) [reply near/omni-token/src/lib.rs]: Since we're not upgrading old tokens, state changes should be fine

File: near/omni-bridge/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (frolvanya) [reply near/omni-bridge/src/lib.rs]: Because we didn't migrated for the ethereum, because we had issues with roles and there wasn't enough reasons to resolve them back then probably

File: near/omni-tests/src/environment.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (kiseln) [reply near/omni-tests/src/environment.rs]: I see the opposite. Bridge contracts needs to do a deposit in order to hold tokens.

File: near/omni-token/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (karim-en) [reply near/omni-token/src/lib.rs]: At some point we will need to update old tokens, we already have 3 different versions (rainbow, omni, poa), it is not a good idea to have 4th one.

File: near/token-deployer/src/lib.rs [hunk 10]
```diff
@@ -50,34 +58,27 @@ impl TokenDeployer {
         contract
     }
 
-    #[payable]
     #[access_control_any(roles(Role::Controller))]
```
[the remark this answers is not in the document]
Comment (frolvanya) [reply near/token-deployer/src/lib.rs]: Fixed in 4d1ffcbf83b1bbdd057446ee154abe8d54b392a1

**Review by karim-en — APPROVED** [review]

**Comment by karim-en** [comment]
@frolvanya could you please fix the conflicts?

**Review by karim-en — APPROVED** [review]
